File: sdk_src/UDP_Client/MobileIMSDK4i_udp_open/MobileIMSDK4i/Core/ClientCoreSDK.m
//  ----------------------------------------------------------------------
//  Copyright (C) 2021  即时通讯网(52im.net) & Jack Jiang.
//  The MobileIMSDK_UDP (MobileIMSDK v6.x UDP版) Project.
//  All rights reserved.
//
//  > Github地址: https://github.com/JackJiang2011/MobileIMSDK
//  > 文档地址:    http://www.52im.net/forum-89-1.html
//  > 技术社区：   http://www.52im.net/
//  > 技术交流群： 215477170 (http://www.52im.net/topic-qqgroup.html)
//  > 作者公众号： “即时通讯技术圈】”，欢迎关注！
//  > 联系作者：   http://www.52im.net/thread-2792-1-1.html
//
//  "即时通讯网(52im.net) - 即时通讯开发者社区!" 推荐开源工程。
//  ----------------------------------------------------------------------

#import "ClientCoreSDK.h"
#import "ChatMessageEvent.h"
#import "ChatBaseEvent.h"
#import "MessageQoSEvent.h"
#import "MBReachability.h"
#import "QoS4SendDaemon.h"
#import "KeepAliveDaemon.h"
#import "LocalDataReciever.h"
#import "LocalSocketProvider.h"
#import "QoS4ReciveDaemon.h"
#import "AutoReLoginDaemon.h"


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
#pragma mark - 静态全局类变量
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

static BOOL ENABLED_DEBUG = NO;
static BOOL autoReLogin = YES;


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
#pragma mark - 私有API
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
@interface ClientCoreSDK ()

@property (nonatomic) BOOL _init;
@property (nonatomic) MBReachability *internetReachability;

@end


////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
#pragma mark - 本类的代码实现
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
@implementation ClientCoreSDK

static ClientCoreSDK *instance = nil;

//------------------------------------------------------
#pragma mark - 静态方法

+ (ClientCoreSDK *)sharedInstance
{
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[super allocWithZone:NULL] init];
    });
    return instance;
}

+ (BOOL) isENABLED_DEBUG
{
    return ENABLED_DEBUG;
}
+ (void) setENABLED_DEBUG:(BOOL)enabledDebug
{
    ENABLED_DEBUG = enabledDebug;
}

+ (BOOL) isAutoReLogin
{
    return autoReLogin;
}
+ (void) setAutoReLogin:(BOOL)arl
{
    autoReLogin = arl;
}


//------------------------------------------------------
#pragma mark - 实例方法

- (id)init
{
    if (![super init])
        return nil;
    
    return self;
}

- (void)initCore
{
    if(!self._init)
    {
        self.connectedToServer = NO;
        self.loginHasInit = NO;
        
        if(self.internetReachability == nil)
        {
            [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(reachabilityChanged:) name:kReachabilityChangedNotification object:nil];
            self.internetReachability = [MBReachability reachabilityForInternetConnection];
        }
        [self.internetReachability startNotifier];
        self._init = YES;
        
        NSLog(@"ClientCoreSDK已经完成initCore了！");
    }
}

- (void) releaseCore
{
    [[AutoReLoginDaemon sharedInstance] stop]; 
    [[QoS4SendDaemon sharedInstance] stop];
    [[KeepAliveDaemon sharedInstance] stop];
    [[QoS4ReciveDaemon sharedInstance] stop];
    [[LocalSocketProvider sharedInstance] closeLocalSocket];

    [[QoS4SendDaemon sharedInstance] clear];
    [[QoS4ReciveDaemon sharedInstance] clear];

    [self.internetReachability stopNotifier];
    
    self._init = NO;
    self.loginHasInit = NO;
    self.connectedToServer = NO;
}


- (BOOL) isInitialed
{
    return self._init;
}

- (void) saveFirstLoginTime:(long)firstLoginTime
{
    if(self.currentLoginInfo != nil)
        self.currentLoginInfo.firstLoginTime = firstLoginTime;
}

// @deprecated 本方法已弃用。请使用 [ClientCoreSDK sharedInstance].currentLoginInfo.loginUserId 替代之！
- (NSString *) currentLoginUserId
{
    return self.currentLoginInfo.loginUserId;
}

- (BOOL)internetReachable
{
    NetworkStatus netStatus = [self.internetReachability currentReachabilityStatus];
    return netStatus == ReachableViaWWAN || netStatus == ReachableViaWiFi;
}

- (void) reachabilityChanged:(NSNotification *)note
{
    MBReachability* reachability = [note object];
    NSParameterAssert([reachability isKindOfClass:[MBReachability class]]);
    
    NetworkStatus netStatus = [reachability currentReachabilityStatus];
    BOOL connectionRequired = [reachability connectionRequired];
    NSString* statusString = @"";
    
    switch (netStatus)
    {
        case NotReachable:
        {
            statusString = NSLocalizedString(@"【IMCORE-UDP】【本地网络通知】检测本地网络连接断开了!", @"Text field text for access is not available");
            connectionRequired = NO;
            
//          self.localDeviceNetworkOk = false;
            [[LocalSocketProvider sharedInstance] closeLocalSocket];
            
            break;
        }
            
        case ReachableViaWWAN: // 蜂窝网络、3G网络等
        case ReachableViaWiFi: // WIFI
        {
            int wifi = (netStatus == ReachableViaWiFi);
            statusString= [NSString stringWithFormat:NSLocalizedString(@"【IMCORE-UDP】【本地网络通知】检测本地网络已连接上了! WIFI? %d", @""), wifi?@"YES":@"NO"];
            
//          self.localDeviceNetworkOk = true;
            [[LocalSocketProvider sharedInstance] closeLocalSocket];
            
            break;
        }
    }
    
    if (connectionRequired)
    {
        NSString *connectionRequiredFormatString = NSLocalizedString(@"【IMCORE-UDP】%@, Connection Required", @"Concatenation of status string with connection requirement");
        statusString = [NSString stringWithFormat:connectionRequiredFormatString, statusString];
    }
    
    if(ENABLED_DEBUG)
        NSLog(@"%@", statusString);
}

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self name:kReachabilityChangedNotification object:nil];
}

@end

File: sdk_src/TCP_Client/MobileIMSDK4j_tcp_Open/src/net/x52im/mobileimsdk/java/ClientCoreSDK.java
/*
 * Copyright (C) 2023  即时通讯网(52im.net) & Jack Jiang.
 * The MobileIMSDK_TCP (MobileIMSDK v6.4 TCP版) Project. 
 * All rights reserved.
 * 
 * > Github地址：https://github.com/JackJiang2011/MobileIMSDK
 * > 文档地址：  http://www.52im.net/forum-89-1.html
 * > 技术社区：  http://www.52im.net/
 * > 技术交流群：215477170 (http://www.52im.net/topic-qqgroup.html)
 * > 作者公众号：“即时通讯技术圈】”，欢迎关注！
 * > 联系作者：  http://www.52im.net/thread-2792-1-1.html
 *  
 * "即时通讯网(52im.net) - 即时通讯开发者社区!" 推荐开源工程。
 * 
 * ClientCoreSDK.java at 2023-9-21 15:32:53, code by Jack Jiang.
 */
package net.x52im.mobileimsdk.java;

import net.x52im.mobileimsdk.java.core.AutoReLoginDaemon;
import net.x52im.mobileimsdk.java.core.KeepAliveDaemon;
import net.x52im.mobileimsdk.java.core.LocalSocketProvider;
import net.x52im.mobileimsdk.java.core.QoS4ReciveDaemon;
import net.x52im.mobileimsdk.java.core.QoS4SendDaemon;
import net.x52im.mobileimsdk.java.event.ChatBaseEvent;
import net.x52im.mobileimsdk.java.event.ChatMessageEvent;
import net.x52im.mobileimsdk.java.event.MessageQoSEvent;
import net.x52im.mobileimsdk.server.protocal.c.PLoginInfo;

public class ClientCoreSDK {
	
	private final static String TAG = ClientCoreSDK.class.getSimpleName();

	public static boolean DEBUG = true;
	public static boolean autoReLogin = true;
	private static ClientCoreSDK instance = null;

	private boolean _init = false;
	private boolean connectedToServer = true;
	private boolean loginHasInit = false;
	private PLoginInfo currentLoginInfo = null;

	private ChatBaseEvent chatBaseEvent = null;
	private ChatMessageEvent chatMessageEvent = null;
	private MessageQoSEvent messageQoSEvent = null;

	public static ClientCoreSDK getInstance() {
		if (instance == null) {
			synchronized (ClientCoreSDK.class) {
				if (instance == null) {
					instance = new ClientCoreSDK();
				}
			}
		}
		return instance;
	}

	private ClientCoreSDK() {

	}

	public void init() {
		if (!_init) {
			_init = true;
		}
	}

	public void release() {
		this.setConnectedToServer(false);
				
		LocalSocketProvider.getInstance().closeLocalSocket();
		AutoReLoginDaemon.getInstance().stop();
		QoS4SendDaemon.getInstance().stop();
		KeepAliveDaemon.getInstance().stop();

		QoS4ReciveDaemon.getInstance().stop();

		QoS4SendDaemon.getInstance().clear();
		QoS4ReciveDaemon.getInstance().clear();

		_init = false;
		this.setLoginHasInit(false);
//		this.setConnectedToServer(false);
	}

	public void setCurrentLoginInfo(PLoginInfo currentLoginInfo)
	{
		this.currentLoginInfo = currentLoginInfo;
	}

	public PLoginInfo getCurrentLoginInfo()
	{
		return this.currentLoginInfo;
	}

	public void saveFirstLoginTime(long firstLoginTime)
	{
		if(this.currentLoginInfo != null)
			this.currentLoginInfo.setFirstLoginTime(firstLoginTime);
	}

	@Deprecated
	public String getCurrentLoginUserId()
	{
		return this.currentLoginInfo.getLoginUserId();
	}

	@Deprecated
	public String getCurrentLoginToken()
	{
		return this.currentLoginInfo.getLoginToken();
	}

	@Deprecated
	public String getCurrentLoginExtra()
	{
		return this.currentLoginInfo.getExtra();
	}

	public boolean isLoginHasInit()
	{
		return loginHasInit;
	}

	public ClientCoreSDK setLoginHasInit(boolean loginHasInit) {
		this.loginHasInit = loginHasInit;
		return this;
	}

	public boolean isConnectedToServer() {
		return connectedToServer;
	}

	public void setConnectedToServer(boolean connectedToServer) {
		this.connectedToServer = connectedToServer;
	}

	public boolean isInitialed() {
		return this._init;
	}

	public void setChatBaseEvent(ChatBaseEvent chatBaseEvent) {
		this.chatBaseEvent = chatBaseEvent;
	}

	public ChatBaseEvent getChatBaseEvent() {
		return chatBaseEvent;
	}

	public void setChatMessageEvent(ChatMessageEvent chatMessageEvent) {
		this.chatMessageEvent = chatMessageEvent;
	}

	public ChatMessageEvent getChatMessageEvent() {
		return chatMessageEvent;
	}

	public void setMessageQoSEvent(MessageQoSEvent messageQoSEvent) {
		this.messageQoSEvent = messageQoSEvent;
	}

	public MessageQoSEvent getMessageQoSEvent() {
		return messageQoSEvent;
	}
}
